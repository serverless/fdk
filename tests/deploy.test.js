const fdk = require('../lib/index')
const eventGatewayProcesses = require('./event-gateway/processes')

const step1 = {
  functions: [
    {
      functionId: 'myFunctionOne',
      provider: {
        type: 'awslambda',
        arn: 'https://xxxxxxxxxx.execute-api.us-east-1.amazonaws.com/dev/test',
        region: 'us-east-1',
      },
    },
    {
      functionId: 'myFunctionTwo',
      provider: {
        type: 'http',
        url: 'http://www.example.com',
      },
    },
  ],
  subscriptions: [
    { functionId: 'myFunctionOne', event: 'pageVisited' },
    { functionId: 'myFunctionOne', event: 'userCreated' },
  ],
}

const step2 = {
  functions: [
    {
      functionId: 'myFunctionOne',
      provider: {
        type: 'awslambda',
        arn: 'https://xxxxxxxxxx.execute-api.us-east-1.amazonaws.com/dev/test',
        region: 'us-east-1',
      },
    },
    {
      functionId: 'myFunctionTwo',
      provider: {
        type: 'http',
        url: 'http://www.example.com',
      },
    },
    {
      functionId: 'myFunctionThree',
      provider: {
        type: 'http',
        url: 'http://www.google.com',
      },
    },
  ],
  subscriptions: [
    { functionId: 'myFunctionOne', event: 'pageVisited' },
    { functionId: 'myFunctionOne', event: 'userCreated' },
    { functionId: 'myFunctionTwo', event: 'userDeleted' },
    { functionId: 'myFunctionThree', event: 'userUpdated' },
  ],
}

const step3 = {
  functions: [],
  subscriptions: [],
}

let eventGateway
let eventGatewayProcessId

beforeAll(() =>
  eventGatewayProcesses
    .spawn({
      configPort: 4017,
      apiPort: 4018,
    })
    .then(processInfo => {
      eventGatewayProcessId = processInfo.id
      eventGateway = fdk.eventGateway({
        url: `http://localhost:${processInfo.apiPort}`,
        configurationUrl: `http://localhost:${processInfo.configPort}`,
      })
    })
)

afterAll(() => {
  eventGatewayProcesses.shutDown(eventGatewayProcessId)
})

test('should created multiple functions and subscriptions if there are none', () => {
  expect.assertions(1)
  return eventGateway
    .deploy(step1)
    .then(response => { expect(response).toMatchSnapshot() })
})

test('should create added functions and subscriptions', () => {
  expect.assertions(2)
  return eventGateway
    .deploy(step1)
    .then(response => { expect(response).toMatchSnapshot('step1') })
    .then(() => eventGateway.deploy(step2))
    .then(response => { expect(response).toMatchSnapshot('step2') })
})

test('should delete functions and subscriptions', () => {
  expect.assertions(2)
  return eventGateway
    .deploy(step1)
    .then(response => { expect(response).toMatchSnapshot('step1') })
    .then(() => eventGateway.deploy(step3))
    .then(response => { expect(response).toMatchSnapshot('step3') })
})
