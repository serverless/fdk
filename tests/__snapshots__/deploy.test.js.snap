// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should created multiple functions and subscriptions if there are none 1`] = `
Object {
  "functions": Array [
    Object {
      "functionId": "myFunctionOne",
      "provider": Object {
        "arn": "https://xxxxxxxxxx.execute-api.us-east-1.amazonaws.com/dev/test",
        "region": "us-east-1",
        "type": "awslambda",
      },
    },
    Object {
      "functionId": "myFunctionTwo",
      "provider": Object {
        "type": "http",
        "url": "http://www.example.com",
      },
    },
  ],
  "subscriptions": Array [
    Object {
      "event": "pageVisited",
      "functionId": "myFunctionOne",
      "subscriptionId": "pageVisited-myFunctionOne",
    },
    Object {
      "event": "userCreated",
      "functionId": "myFunctionOne",
      "subscriptionId": "userCreated-myFunctionOne",
    },
  ],
}
`;

exports[`step1 1`] = `
Object {
  "functions": Array [],
  "subscriptions": Array [
    Object {
      "event": "pageVisited",
      "functionId": "myFunctionOne",
      "subscriptionId": "pageVisited-myFunctionOne",
    },
    Object {
      "event": "userCreated",
      "functionId": "myFunctionOne",
      "subscriptionId": "userCreated-myFunctionOne",
    },
  ],
}
`;

exports[`step1 2`] = `
Object {
  "functions": Array [],
  "subscriptions": Array [
    Object {
      "event": "pageVisited",
      "functionId": "myFunctionOne",
      "subscriptionId": "pageVisited-myFunctionOne",
    },
    Object {
      "event": "userCreated",
      "functionId": "myFunctionOne",
      "subscriptionId": "userCreated-myFunctionOne",
    },
  ],
}
`;

exports[`step2 1`] = `
Object {
  "functions": Array [
    Object {
      "functionId": "myFunctionThree",
      "provider": Object {
        "type": "http",
        "url": "http://www.google.com",
      },
    },
  ],
  "subscriptions": Array [
    Object {
      "event": "pageVisited",
      "functionId": "myFunctionOne",
      "subscriptionId": "pageVisited-myFunctionOne",
    },
    Object {
      "event": "userCreated",
      "functionId": "myFunctionOne",
      "subscriptionId": "userCreated-myFunctionOne",
    },
    Object {
      "event": "userDeleted",
      "functionId": "myFunctionTwo",
      "subscriptionId": "userDeleted-myFunctionTwo",
    },
    Object {
      "event": "userUpdated",
      "functionId": "myFunctionThree",
      "subscriptionId": "userUpdated-myFunctionThree",
    },
  ],
}
`;

exports[`step3 1`] = `
Object {
  "functions": Array [],
  "subscriptions": Array [],
}
`;
